engine/src/vsfilesystem.cpp

```diff
@@ -741,51 +742,102 @@ void InitMods() {
         selectcurrentdir = datadir + "/..";
-        int ret = scandir(selectcurrentdir.c_str(), &dirlist, selectdirs, 0);
-        if (ret >= 0) {
-            while (ret--) {
-                string dname(dirlist[ret]->d_name);
-                if (dname == game_options()->hqtextures) {
-                    curpath = selectcurrentdir + "/" + dname;
-                    VS_LOG(info, "\n\nAdding HQ Textures Pack\n\n");
-                    Rootdir.push_back(curpath);
-                }
+        boost::filesystem::path p(selectcurrentdir);
+        for (boost::filesystem::directory_entry& entry : boost::filesystem::directory_iterator(p)) {
+            const boost::filesystem::path& filename = entry.path().filename();
+            const std::string filename_string = filename.string();
+            if (filename_string == game_options()->hqtextures && is_directory(entry.status())) {
+                curpath.clear();
+                curpath.append(selectcurrentdir);
+                curpath.append("/");
+                curpath.append(filename_string);
+                VS_LOG(important_info, "\n\nAdding HQ Textures Pack\n\n");
+                Rootdir.push_back(curpath);
             }
         }
-        free(dirlist);
+
+//        int ret = scandir(selectcurrentdir.c_str(), &dirlist, selectdirs, nullptr);
+//        if (ret >= 0) {
+//            while (ret--) {
+//                string dname(dirlist[ret]->d_name);
+//                if (dname == game_options()->hqtextures) {
+//                    curpath.clear();
+//                    curpath.append(selectcurrentdir);
+//                    curpath.append("/");
+//                    curpath.append(dname);
+//                    VS_LOG(info, "\n\nAdding HQ Textures Pack\n\n");
+//                    Rootdir.push_back(curpath);
+//                }
+//            }
+//        }
+//        free(dirlist);
     }
+
     selectcurrentdir = moddir;
-    int ret = scandir(selectcurrentdir.c_str(), &dirlist, selectdirs, 0);
-    if (ret < 0) {
-        return;
-    } else {
-        while (ret--) {
-            string dname(dirlist[ret]->d_name);
-            if (dname == modname) {
-                curpath = moddir + "/" + dname;
-                VS_LOG(info, (boost::format("Adding mod path : %1%") % curpath));
+    boost::filesystem::path mod_path(selectcurrentdir);
+    if (exists(mod_path)) {
+        for (boost::filesystem::directory_entry &entry : boost::filesystem::directory_iterator(mod_path)) {
+            const boost::filesystem::path &filename = entry.path().filename();
+            const std::string filename_string = filename.string();
+            if (filename_string == modname && is_directory(entry.status())) {
+                curpath.clear();
+                curpath.append(selectcurrentdir);
+                curpath.append("/");
+                curpath.append(filename_string);
+                VS_LOG(important_info, (boost::format("Adding mod path : %1%") % curpath));
```

Comment:
Seems like we could abstract this one layer and DRY the code a little.
I do like the change, BTW.